### PR TITLE
ICU-21289 Switch to using CharString for calling uloc_getKeywordValue()

### PR DESCRIPTION
**Pull request by roubert**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21289
- [X] Required: The PR title must be prefixed with a JIRA Issue number.
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number.
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/common/ucurr.cpp [hunk 1]
```diff
@@ -520,14 +521,18 @@ ucurr_forLocale(const char* locale,
...
+        ulocimp_getKeywordValue(locale, "currency", sink, &localStatus);
+    }
+    int32_t resLen = currency.length();
+
+    if (U_SUCCESS(localStatus) && resLen == 3 && uprv_isInvariantString(currency.data(), resLen)) {
         if (resLen < buffCapacity) {
-            T_CString_toUpperCase(currency);
-            u_charsToUChars(currency, buff, resLen);
+            T_CString_toUpperCase(currency.data());
+            u_charsToUChars(currency.data(), buff, resLen);
```
Comment (markusicu) [inline icu4c/source/common/ucurr.cpp:535]: Consider using currency.appendInvariantChars(buff, resLen, ec);

**Review by markusicu — COMMENTED** [review]
thanks, good changes!

consider timing vs. ICU 74:
- code freeze is in less than a week
- i assume that you are planning further changes like this
- if you get this in now for 74, then changes after the code freeze need to go into 75 under a new ticket split off from this one

File: icu4c/source/common/ucurr.cpp (hunk 1 above)
Comment (roubert) [reply icu4c/source/common/ucurr.cpp:535]: Done.

**Comment by roubert** [comment]
> * code freeze is in less than a week

Yes, so I'm just doing some clean-up before the freeze.

> * i assume that you are planning further changes like this

Definitely.

> * if you get this in now for 74, then changes after the code freeze need to go into 75 under a new ticket split off from this one

As already so many times before.

File: icu4c/source/common/ucurr.cpp (hunk 1 above)
Comment (roubert) [reply icu4c/source/common/ucurr.cpp:535]: You read this backwards, right? The existing code copies from `currency` to `buff`. Calling `appendInvariantChars()` instead would copy in the opposite direction.

Comment (markusicu) [reply icu4c/source/common/ucurr.cpp:535]: oops! sorry!

**Comment by markusicu** [comment]
> > * if you get this in now for 74, then changes after the code freeze need to go into 75 under a new ticket split off from this one
> 
> As already so many times before.

Possible option: Create a new ticket specifically for the 74 changes (using this PR title for the ticket title), change this PR to use that, and punt the existing ticket to 75.

**Comment by roubert** [comment]
> We wouldn't need to split an open-ended ticket into two open-ended tickets.

But the nature of this work is that it's open-ended, there's no natural or logical boundary for what'll go into this and into the next release, it'll be two open-ended tickets regardless of what we do.

**Comment by markusicu** [comment]
> > We wouldn't need to split an open-ended ticket into two open-ended tickets.
> 
> But the nature of this work is that it's open-ended, there's no natural or logical boundary for what'll go into this and into the next release, it'll be two open-ended tickets regardless of what we do.

We can do it that way, but because we have just two days or so left for ICU 74, work going into this release is not open-ended. Just thinking that we could carve off a specific ticket for the 74 work.

**Comment by roubert** [comment]
> We can do it that way, but because we have just two days or so left for ICU 74, work going into this release is not open-ended. Just thinking that we could carve off a specific ticket for the 74 work.

That's what I mean by open-ended: The end is defined by when we run out of time, not at any particular set of changes that could be enumerated.

**Review by markusicu — APPROVED** [review]

**Pull request merged**